plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "openssl.hpp"
+
+#include "openssl/err.h"
+#include "openssl/evp.h"
+#include "openssl/pem.h"
+
+#include "jlog.hpp"
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+#define CURRENT_OPENSSL_VERSION "openssl plugin 1.0.0"
+
+static const char* opensslCompatibleVersions[] = {
+    CURRENT_OPENSSL_VERSION,
+    NULL };
+
+OPENSSL_API bool OPENSSL_CALL getECLPluginDefinition(ECLPluginDefinitionBlock* pb)
+{
+    if (pb->size == sizeof(ECLPluginDefinitionBlockEx))
+    {
+        ECLPluginDefinitionBlockEx* pbx = static_cast<ECLPluginDefinitionBlockEx*>(pb);
+        pbx->compatibleVersions = opensslCompatibleVersions;
+    }
+    else if (pb->size != sizeof(ECLPluginDefinitionBlock))
+    {
+        return false;
+    }
+
+    pb->magicVersion = PLUGIN_VERSION;
+    pb->version = CURRENT_OPENSSL_VERSION;
+    pb->moduleName = "openssl";
+    pb->ECL = NULL;
+    pb->flags = PLUGIN_IMPLICIT_MODULE;
+    pb->description = "ECL plugin library for the C++ API in OpenSSL";
+
+    return true;
+}
+
+namespace nsOpenSSL
+{
+
+void failOpenSSLError(const std::string& context)
+{
+    unsigned long   errCode = 0;
+    char            buffer[120];
+
+    ERR_error_string_n(ERR_get_error(), buffer, sizeof(buffer));
+
+    std::string desc = "Error within ";
+    desc += context;
+    desc += ": ";
+    desc += buffer;
+
+    rtlFail(-1, desc.c_str());
+}
+
+//called during PEM_read_bio_PrivateKey to set passphrase
+int passphraseCB(char *passPhraseBuf, int passPhraseBufSize, int rwflag, void *pPassPhraseMB)
+{
+    size32_t len = ((MemoryBuffer*)pPassPhraseMB)->length();
+    if (passPhraseBufSize >= (int)len)
+    {
+        memcpy(passPhraseBuf, ((MemoryBuffer*)pPassPhraseMB)->bufferBase(), len);
+        return len;
+    }
+    PROGLOG("Private Key Passphrase too long (%d bytes), max %d", len, passPhraseBufSize);
+    return 0;
+}
+
+bool isPublicKey(size32_t keyLen, const char * key)
+{
+    for (int i = 0; key[i] != '\n' && i < (keyLen - 15); i++)
+    {
+        if (key[i] == 'P')
+        {
+            if (strncmp(key + i + 1, "UBLIC KEY-----", 14) == 0)
+                return true;
+        }
+    }
+    return false;
+}
+
+static constexpr int OPENSSL_MAX_CACHE_SIZE = 10;
+static constexpr bool PRINT_STATS = false;
+template <typename T>
+class OpenSSLCache
+{
+public:
+    const T * checkCache(const char * algorithm_name)
+    {
+        for (auto& c : cache)
+        {
+            if (std::get<0>(c) == algorithm_name)
+            {
+                hits++;
+                return std::get<1>(c);
+            }
+        }
+        misses++;
+        const T * newObj = getObjectByName(algorithm_name);
```

Comment:
I think these will never be freed.  Should this be using a unique_ptr in the tuple?